File: calc.hs
#!/usr/bin/env runhaskell

import Control.Monad (foldM, guard)
import System.IO (isEOF)
import Text.Read (readMaybe)

process stack token =
  case readMaybe token of
    Just x  -> Right $ x : stack
    Nothing -> readOp token >>= evalOpOnStack
  where
    readOp t = case t of
      "+" -> Right (+)
      "-" -> Right (-)
      "*" -> Right (*)
      "/" -> Right quot
      _   -> Left "Invalid Token"

    evalOpOnStack op = case stack of
      (b : a : rest) -> Right $ op a b : rest
      _              -> Left "Too Few Arguments"

eval line = do
  stack <- foldM process [] (words line)
  case stack of
    [x] -> Right x
    _   -> Left "Invalid Input"

main = do
  isEOF >>= guard . not
  line <- getLine
  case eval line of
    Left e  -> putStrLn e
    Right x -> print x
  main

File: calc.lua
#!/usr/bin/env lua

while true do
	local stack = {};

	for t in io.read("*l"):gmatch("%S+") do
		if t:match("^%s*-?%d+%s*$") then
			stack[#stack + 1] = math.tointeger(t)
			goto continue
		end

		if #stack < 2 then
			print("Too Few Arguments")
			goto continue_outer
		end

		local b = table.remove(stack)
		local a = table.remove(stack)

		if t == "+" then stack[#stack + 1] = a + b
		elseif t == "-" then stack[#stack + 1] = a - b
		elseif t == "*" then stack[#stack + 1] = a * b
		elseif t == "/" then stack[#stack + 1] = a / b
		else
			print("Invalid Token")
			goto continue_outer
		end

		::continue::
	end

	if #stack == 1 then
		print(stack[1])
	else
		print("Invalid Input")
	end
	
	::continue_outer::
end
